Add Rake::Task#prerequisite_tasks!

jasoares opened the pull request:
Add `Rake::Task#prerequisite_tasks!` method to return a full list of prerequisite tasks from the receiver tasks and its prerequisite tasks tasks, recursively.

This is usefull if one needs to call reenable on a task and all of its prerequisite tasks, not just direct ones.
As a use case example, when writing integration tests for rake tasks it is common to call `Rake::Task#reenable` before or after every `Rake::Task#invoke`.
If the task has prerequisites they will not be invoked beyond the first call because they are not affected by #reenable. Calling #reenable on every task's prerequisite task can be done with just `Rake::Task[task].prerequisite_tasks.each(&:reenable)`
but that is not so simple when those prerequisites have themselves prerequisite tasks, that's when this comes in handy.
I hope the example is clear enough.

Other alternatives for the method name may be `#deep_prerequisite_tasks` based on `#deep_clone` or `#deep_copy`.

The implementation proposed sets a default depth limit of 30 to avoid circular dependency(SystemStackError: stack level too deep).

jasoares committed:
Add Rake::Task#prerequisite_tasks!

Add Rake::Task#prerequisite_tasks! method to return a full list
of prerequisite tasks from the receiver tasks and its prerequisite
tasks tasks, recursively.

This is usefull if one needs to call reenable on a task and all
of its prerequisite tasks.

Other alternatives for the method name may be #deep_prerequisite_tasks
based on deep_clone or deep_copy.

The implementation proposed sets a default depth limit of 30 to avoid
circular dependency(SystemStackError: stack level too deep).

Signed-off-by: João Soares <jsoaresgeral@gmail.com>

jasoares committed:
Add comments to Rake::Task#prerequisite_tasks!

Signed-off-by: João Soares <jsoaresgeral@gmail.com>

jimweirich commented:
Ok, I can see a use for this.  A couple of issues.

(a) Why does the name end in a bang?  I don't see the method as being particularly dangerous in any way. I think a name like "all_prerequisite_tasks" would be better.

(b) The method should not return duplicates.  By doing so will automatically make it safe for cyclic dependencies, so the depth argument can be dropped.

(c) The ordering of the returned prerequisites should not be specified by the tests. Add a .sort_by { |t| t.name } to the return value before comparing for equality.

jasoares committed:
New implementation without duplicates

Rename Rake::Task#prerequisite_tasks! to
Rake::Task#all_prerequisite_tasks.

Reimplement Rake::Task#all_prerequisite_tasks to not return duplicates.

Add protected helper method to recursively collect dependencies
rejecting duplicates.

Abstract tests from returned list order.

Signed-off-by: João Soares <jsoaresgeral@gmail.com>

jasoares commented:
Let me know if everything is ok...

jimweirich commented:
Looks good.
